src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php

```diff
@@ -27,11 +33,55 @@ class EntityType extends DoctrineType
      */
     public function getLoader(ObjectManager $manager, $queryBuilder, $class)
     {
-        return new ORMQueryBuilderLoader(
-            $queryBuilder,
-            $manager,
-            $class
-        );
+        if (!$queryBuilder instanceof QueryBuilder) {
+            return new ORMQueryBuilderLoader(
+                $queryBuilder,
+                $manager,
+                $class
+            );
+        }
+
+        $queryBuilderHash = $this->getQueryBuilderHash($queryBuilder);
+        $loaderHash = $this->getLoaderHash($manager, $queryBuilderHash, $class);
+
+        if (!isset($this->loaderCache[$loaderHash])) {
+            $this->loaderCache[$loaderHash] = new ORMQueryBuilderLoader(
+                $queryBuilder,
+                $manager,
+                $class
+            );
+        }
+
+        return $this->loaderCache[$loaderHash];
+    }
+
+    /**
+     * @param QueryBuilder $queryBuilder
+     *
+     * @return string
+     */
+    private function getQueryBuilderHash(QueryBuilder $queryBuilder)
+    {
+        return hash('sha256', json_encode(array(
+            'sql' => $queryBuilder->getQuery()->getSQL(),
+            'parameters' => $queryBuilder->getParameters(),
```

Comment:
parameters can contain objects (entities or DateTime objects for instance), which will not work well here